src/ophyd_async/epics/pmac/_pmacTrajectory.py

```diff
@@ -0,0 +1,154 @@
+import time
+
+from bluesky.protocols import Flyable, Preparable
+from scanspec.specs import Frames
+
+from ophyd_async.core.async_status import AsyncStatus, WatchableAsyncStatus
+from ophyd_async.core.signal import observe_value
+from ophyd_async.core.utils import WatcherUpdate
+from ophyd_async.epics.motion import Motor
+from ophyd_async.epics.pmac import Pmac
+
+TICK_S = 0.000001
+
+
+class PmacTrajectory(Pmac, Flyable, Preparable):
+    """Device that moves a PMAC Motor record"""
+
+    def __init__(self, prefix: str, cs: int, name="") -> None:
+        # Make a dict of which motors are for which cs axis
+        self._fly_start: float
+        self.cs = cs
+        super().__init__(prefix, cs, name=name)
+
+    async def _ramp_up_velocity_pos(
+        self, velocity: float, motor: Motor, end_velocity: float
+    ):
+        # Assuming ramping to or from 0
+        max_velocity_acceleration_time = await motor.acceleration_time.get_value()
+        max_velocity = await motor.max_velocity.get_value()
+        delta_v = abs(end_velocity - velocity)
+        accl_time = max_velocity_acceleration_time * delta_v / max_velocity
+        disp = 0.5 * (velocity + end_velocity) * accl_time
+        return [disp, accl_time]
+
+    @AsyncStatus.wrap
+    async def prepare(self, stack: list[Frames[Motor]]):
+        # Which Axes are in use?
+
+        scanSize = len(stack[0])
+        scanAxes = stack[0].axes()
+
+        cs_ports = set()
+        self.profile = {}
+        cs_axes: dict[Motor, str] = {}
+        for axis in scanAxes:
+            if axis != "DURATION":
+                cs_port, cs_axis = await self.get_cs_info(axis)
+                self.profile[cs_axis] = []
+                self.profile[cs_axis + "_velocity"] = []
+                cs_ports.add(cs_port)
+                cs_axes[axis] = cs_axis
+            else:
+                self.profile[axis.lower()] = []
+        assert len(cs_ports) == 1, "Motors in more than one CS"
+        cs_port = cs_ports.pop()
+
+        # Calc Velocity
+
+        for axis in scanAxes:
+            for i in range(scanSize):
+                if axis != "DURATION":
+                    self.profile[cs_axes[axis] + "_velocity"].append(
+                        (stack[0].upper[axis][i] - stack[0].lower[axis][i])
+                        / (stack[0].midpoints["DURATION"][i])
+                    )
+                    self.profile[cs_axes[axis]].append(stack[0].midpoints[axis][i])
+                else:
+                    self.profile[axis.lower()].append(
+                        int(stack[0].midpoints[axis][i] / TICK_S)
+                    )
```

Comment:
Eventually we should make these numpy operations so we get some speed, but this will do for now.

We also need to make sure that we put lower and midpoint points in here, but that can also be a point for later.